common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -447,6 +447,37 @@ private UTF8String toUpperCaseSlow() {
     return fromString(toString().toUpperCase());
   }
 
+  /**
+   * Optimized lowercase comparison for UTF8_BINARY_LCASE collation
+   */
+  public int compareLowercase(UTF8String other) {
+    int curr;
+    for (curr = 0; curr < numBytes && curr < other.numBytes; curr++) {
+      byte left = getByte(curr);
+      byte right = other.getByte(curr);
+      if (numBytesForFirstByte(left) != 1 || numBytesForFirstByte(right) != 1) {
+        return compareLowercaseSuffixSlow(other, curr);
+      }
+      int lowerLeft = Character.toLowerCase(left);
+      int lowerRight = Character.toLowerCase(right);
+      if (lowerLeft > 127 || lowerRight > 127) {
```

Comment:
I see that you are not introducing anything new here and that `numBytes != 1 && codePoint < 127` is already used in `toUpperCase`. But I don't really understand this logic.
Why can't we take multibyte codepoints? I see that `Character.ToLowerCase` accepts an integer specifying code point, that we can decode from UTF8 binary. What is the reason why we can't use this for any code point?

Reply:
The idea is that if we remain in ASCII space, we don't need to worry about locales and length of character representation in UTF8 encoding; if we step outside this set of characters, we may encounter issues with both. 

Ignoring locales (which `Character.ToLowerCase` does) means that we will break compliance with `UTF8String.toLowerCase` which is locale-dependent. Issue with varying length of character encoding in bytes introduces major performance penalty as we then need to deal with byte buffers to store incoming converted data. Although this method should have same asymptotic complexity, it will be several times slower than simply comparing char-by-char. Since current naive implementation is not that much bad (~7x slower than UTF8_BINARY), we woudn't gain much in terms of performance.